Dark Mode

shrianshChari opened the pull request:
A dark theme is one of the features that is requested in #348, and since it seemed like basic CSS and JS work, I figured I'd give it a go. Here is the result.

### Shortcomings

- Have not changed the styling for the navbar at the top, so it is very bright
- Could not find documentation on changing the styles of the Select2 dropdowns, so they stay light gray
- Because of the timing of which the JavaScript will load, the web page will undergo [FOIT](https://css-tricks.com/flash-of-inaccurate-color-theme-fart/), where the web page will briefly flash as light mode before entering dark mode. Based on what I've read, the main way to ensure this isn't a problem is to utilize server-side rendering to serve the web pages directly in dark mode, which I don't think is what this page is designed to do.
  - I have an idea of how this could potentially be circumvented, but I don't know if it would actually work. Essentially, we would have a defined CSS file (maybe `styles.css`) that would serve as the file that gets used as the style that gets loaded in when the web page loads. Then, the user can select a theme from a dropdown menu, and the page would make the styles corresponding to that theme as the new `styles.css` to pull from, then reload the page to reapply the styles.
  - This would basically be pure JavaScript and would take more time to implement, not to mention I don't entirely know if it would prevent FOIT to begin with.

### Screenshots
<details>
   <summary>Open me!</summary>

![image](https://user-images.githubusercontent.com/30420527/215252034-b7f29944-60b9-47bc-8764-e07e109aad21.png)

![Screenshot from 2023-01-28 00-43-03](https://user-images.githubusercontent.com/30420527/215251212-88103600-08fb-4cd8-9296-1149e8df0cd3.png)

![image](https://user-images.githubusercontent.com/30420527/215251251-433dc089-3d1e-447f-a6c7-b7c4380380e9.png)
</details>

urkerab commented on src/js/dark-theme-toggle.js:
```diff
@@ -0,0 +1,55 @@
+	darkThemeButton.innerText = prefersDarkTheme ? 'Click for Light Theme' : 'Click for Dark Theme';
+}
+
+darkThemeButton.addEventListener('click', function () {
+	// Idk why this can't be directly called, but oh well
+	toggleTheme();
+});
```
Was this what you tried?
```suggestion
darkThemeButton.addEventListener('click', toggleTheme);
```

shrianshChari replied:
I believe I tried it with the parentheses, but I didn't try it without the parentheses. Looking over my own code, I realize that I never actually call the anonymous function that I had defined.

urkerab commented on src/js/dark-theme-toggle.js:
```diff
@@ -0,0 +1,55 @@
+* render.
+*/
+
+var prefersDarkTheme = localStorage.getItem('darkTheme');
+
+/*
+* localStorage will only store strings
+* This means that if it has the value 'false',
+* It will be truey and incorrectly cause the
+* dark theme to load.
+*/
+if (prefersDarkTheme == 'true') {
+	prefersDarkTheme = true;
+} else {
+	prefersDarkTheme = false;
+}
```
```suggestion
/*
 * localStorage will only store strings.
 * This means that if it has the value 'false',
 * it will be truthy and incorrectly cause the
 * dark theme to load.
 */
var prefersDarkTheme = localStorage.getItem('darkTheme') === 'true';
```

urkerab commented on src/index.template.html:
```diff
@@ -7,6 +7,7 @@
     <link type="text/css" rel="stylesheet" href="./js/vendor/select2/select2.css" />
     <link type="text/css" rel="stylesheet" href="./css/main.css?" />
     <link type="text/css" rel="stylesheet" href="./css/vendor/bootstrap.css" />
+    <link rel="stylesheet" type="text/css" href="./css/dark-theme.css">
```
You should switch `type` and `rel` for consistency with the other style sheets.

You might be able to reduce FOIT by just setting the theme class on the `<html>` element (which you can do while the page is loading, although that doesn't help you with setting the button label) and then use descendant selectors for all of your dark mode overrides.

I know this isn't how Pokémon Showdown! does it, but the faster way of doing this is to use a disabled stylesheet; add `disabled="true"` here and also give this element an `id`, and then if the user wants dark mode you can toggle the disabled property of the link. Again, this should also work while the page is still loading to reduce FOIT. (It also avoids needing a separate variable to track whether the user prefers dark mode). You'll also need to remove the dark-theme class on the styles, since they'll be activated automatically depending on whether the stylesheet is disabled. This feature does require IE 9 or later, if that's relevant.

If the content delivery system used by the damage calculator is capable of doing it then using cookies instead of local storage would allow it to send the correct attribute (either `<html class="dark-theme">` or `<stylesheet disabled="true">` as appropriate).

shrianshChari commented:
> I know this isn't how Pokémon Showdown! does it, but the faster way of doing this is to use a disabled stylesheet; add `disabled="true"` here and also give this element an `id`, and then if the user wants dark mode you can toggle the disabled property of the link. Again, this should also work while the page is still loading to reduce FOIT. (It also avoids needing a separate variable to track whether the user prefers dark mode). You'll also need to remove the dark-theme class on the styles, since they'll be activated automatically depending on whether the stylesheet is disabled. This feature does require IE 9 or later, if that's relevant.

I've implemented this functionality into the dark theme, and it does reduce most of the FOIT (though there is still a little bit of FOIT in the background of where the panels are. 

<details>
   <summary>What it looks like</summary>

![image](https://user-images.githubusercontent.com/30420527/215295966-a8da3b46-16d8-400a-a6e4-fa53bb0136c6.png)
</details>

Additionally, I'd like to implement this into the Random Battles calculator (and perhaps the 1vAll and Allv1 calculators, but that would take a good amount of time to add the class names to the HTML elements.

AustinXII commented:
Thank you for this, this looks really good, but I don't want to merge this without having support for all modes. Switching between randoms and the normal calc etc and going back and forth between light and dark will not go well.

shrianshChari commented:
Absolutely fine, I'll update when I've finished adding dark mode support across all the modes.

shrianshChari commented:
Ok, I've added dark mode for the other three calculators. I couldn't figure out how to style some of the stuff in the Honkalculator because it was all inserted using jQuery (namely, there's this little piece of text with a search box on the left that I don't know how to define styles for (it has neither a class nor an id):
<details>
<summary>Secret Search</summary>

![Screenshot from 2023-03-04 20-33-15](https://user-images.githubusercontent.com/30420527/222941541-d56b0275-4595-4b31-bf6d-52c170a73c1f.png)
</details>

Otherwise, I've realized what the "cascading" in CSS means (aka I've removed the `dark-theme` class) and applied it to the other three calculators:

<details>
<summary>Random Battles</summary>

![Screenshot from 2023-03-04 00-59-20](https://user-images.githubusercontent.com/30420527/222941643-7b4d06f0-eeb6-4b34-9fcf-22d41d792f72.png)
![Screenshot from 2023-03-04 00-59-27](https://user-images.githubusercontent.com/30420527/222941645-2ec3df34-7788-425c-ba81-c7c861d8e205.png)
</details>

<details>
<summary>Honkalculator</summary>

![Screenshot from 2023-03-04 00-59-38](https://user-images.githubusercontent.com/30420527/222941697-efc83f39-67cf-4a66-a258-1c88139ec489.png)

![Screenshot from 2023-03-04 01-00-16](https://user-images.githubusercontent.com/30420527/222941702-cfad5e60-7a34-47e3-8685-83af3db19a57.png)

![Screenshot from 2023-03-04 01-00-07](https://user-images.githubusercontent.com/30420527/222941709-9bea1170-fc23-44c5-aabb-82e84f7d103e.png)

</details>

shrianshChari commented:
I figured out a method to style that "Search" label.

<details>
<summary>Fixed Search</summary>

![Screenshot from 2023-03-05 00-03-46](https://user-images.githubusercontent.com/30420527/222942683-e69a153b-f2d4-40c3-839f-c50fda11e9b9.png)


</details>

AustinXII commented:
perfect, thank you!
